mycpp/marksweep_heap.h

```diff
@@ -97,6 +105,9 @@ class RootSet {
   // root_set_[2] is being pushed/popped/modified.
   std::vector<std::vector<Obj*>> stack_;
   int num_frames_ = 0;  // frames 0 to N-1 are valid
+
+  // Track globals separately from the stack to avoid mistakes.
+  std::vector<Obj*> globals_;
```

Comment:
I think we should use `stack_[0]` instead, which will simplify the code.  Notice that we initialize `num_frames_(1)` for this frame.

This frame is currently

- the one that is handled by DoProcessExit(bool) -- it pops the last frame and Collect()
- That is, It's  for objects that are rooted until the end of main() -- i.e. returned by a function called by main

It can also be the "global" frame

I think the same tests will pass, but the code will be simpler

Reply:
Ah ok cool. I thought about using `stack_[0]` but wasn't super confident that we could count on it never being accidentally popped. Will update this and add a test

Reply:
In retrospect, this assert should have been sufficiently convincing... https://github.com/oilshell/oil/blob/master/mycpp/marksweep_heap.h#L74